e2e/tests/interchain_accounts/query_test.go

```diff
@@ -0,0 +1,163 @@
+		// explicitly set the version string because we don't want to use incentivized channels.
+		version := icatypes.NewDefaultMetadataString(ibctesting.FirstConnectionID, ibctesting.FirstConnectionID)
+		msgRegisterAccount := controllertypes.NewMsgRegisterInterchainAccount(ibctesting.FirstConnectionID, controllerAddress, version, channeltypes.UNORDERED)
+
+		txResp := s.BroadcastMessages(ctx, chainA, controllerAccount, msgRegisterAccount)
+		s.AssertTxSuccess(txResp)
+	})
+
+	t.Run("start relayer", func(t *testing.T) {
+		s.StartRelayer(relayer)
+	})
+
+	t.Run("verify interchain account", func(t *testing.T) {
+		var err error
+		hostAccount, err = s.QueryInterchainAccount(ctx, chainA, controllerAddress, ibctesting.FirstConnectionID)
+		s.Require().NoError(err)
+		s.Require().NotZero(len(hostAccount))
+
+		channels, err := relayer.GetChannels(ctx, s.GetRelayerExecReporter(), chainA.Config().ChainID)
+		s.Require().NoError(err)
+		s.Require().Equal(len(channels), 2)
+	})
+
+	t.Run("query via interchain account", func(t *testing.T) {
+		// the host account need not be funded
+		t.Run("broadcast query packet", func(t *testing.T) {
+			balanceQuery := banktypes.NewQueryBalanceRequest(chainBAccount.Address(), chainB.Config().Denom)
+			queryBz, err := balanceQuery.Marshal()
+			s.Require().NoError(err)
+
+			queryMsg := icahosttypes.NewMsgModuleQuerySafe(hostAccount, []*icahosttypes.QueryRequest{
+				{
+					Path: "/cosmos.bank.v1beta1.Query/Balance",
+					Data: queryBz,
+				},
+			})
+
+			cdc := testsuite.Codec()
+			bz, err := icatypes.SerializeCosmosTx(cdc, []proto.Message{queryMsg}, icatypes.EncodingProtobuf)
+			s.Require().NoError(err)
+
+			packetData := icatypes.InterchainAccountPacketData{
+				Type: icatypes.EXECUTE_TX,
+				Data: bz,
+				Memo: "e2e",
+			}
+
+			icaQueryMsg := controllertypes.NewMsgSendTx(controllerAddress, ibctesting.FirstConnectionID, uint64(time.Hour.Nanoseconds()), packetData)
+
+			txResp := s.BroadcastMessages(ctx, chainA, controllerAccount, icaQueryMsg)
+			s.AssertTxSuccess(txResp)
+
+			s.Require().NoError(testutil.WaitForBlocks(ctx, 10, chainA, chainB))
+		})
+
+		t.Run("verify query response", func(t *testing.T) {
+			txSearchRes, err := s.QueryTxsByEvents(ctx, chainB, 1, 1, "message.action='/ibc.core.channel.v1.MsgRecvPacket'", "")
+			s.Require().NoError(err)
+			s.Require().Len(txSearchRes.Txs, 1)
+
+			// get acknowledgement
+			ackFound := false
+			ack := &channeltypes.Acknowledgement_Result{}
+
+		search_ack:
+			for _, event := range txSearchRes.Txs[0].Events {
+				if event.Type != channeltypes.EventTypeWriteAck {
+					continue
+				}
+
+				for _, attr := range event.Attributes {
+					if attr.Key != channeltypes.AttributeKeyAckHex {
+						continue
+					}
+
+					ackBz, err := hex.DecodeString(attr.Value)
+					s.Require().NoError(err)
+
+					err = json.Unmarshal(ackBz, ack)
+					s.Require().NoError(err)
+
+					ackFound = true
+
+					break search_ack
+				}
+			}
```

Comment:
This should probably be its own function.

Reply:
I've already pulled this to its own function. I'll probably make another PR for improving this as I think this function can go to `testing/events.go` as the same logic is also used there many times